lib/dispatch/repositories/github_client.ex

```diff
@@ -84,84 +75,83 @@ defmodule Dispatch.Repositories.GitHubClient do
     }
     """
 
-    body =
-      Poison.encode!(%{
-        query: query,
-        variables: %{
-          repoOwner: repo_owner,
-          repoName: repo_name
-        }
-      })
-
-    (@base_url <> "/graphql")
-    |> HTTPoison.post(body, graphql_headers())
-    |> case do
-      {:ok, %{status_code: 200, body: body}} ->
-        map_requestable_users_response(body)
+    request = %{
+      query: query,
+      variables: %{
+        repoOwner: repo_owner,
+        repoName: repo_name
+      }
+    }
 
+    with {:ok, body} <- Jason.encode(request),
+         {:ok, %{status_code: 200, body: body}} <- HTTPoison.post(@base_url <> "/graphql", body, graphql_headers()) do
+      map_requestable_users(body)
+    else
       _ ->
         :timer.sleep(retry_delay())
         fetch_requestable_users(repo, try_count + 1)
     end
   end
 
-  defp map_requestable_users_response(body) do
-    body
-    |> Poison.decode()
-    |> map_requestable_users()
-  end
+  defp map_requestable_users(body) do
+    case Jason.decode(body) do
+      {:ok, data} when is_map(data) ->
+        data
+        |> get_in(["data", "repository", "assignableUsers", "nodes"])
+        |> Enum.map(fn %{"login" => username, "name" => fullname} -> %User{username: username, fullname: fullname} end)
 
-  defp map_requestable_users({:ok, data}) do
-    data
-    |> get_in(["data", "repository", "assignableUsers", "nodes"])
-    |> Enum.map(fn %{"login" => username, "name" => fullname} -> %User{username: username, fullname: fullname} end)
+      _ ->
+        []
+    end
   end
 
-  defp map_requestable_users(_), do: []
+  def fetch_contributors(repo, try_count \\ 0)
 
-  def fetch_contributors(repo) do
-    fetch_contributors(repo, 0)
+  def fetch_contributors(_, try_count) when try_count > @retry_count do
+    []
   end
 
-  defp fetch_contributors(_, try_count) when try_count > @retry_count, do: []
-
-  defp fetch_contributors(repo, try_count) do
+  def fetch_contributors(repo, try_count) do
     (@base_url <> "/repos/#{repo}/stats/contributors")
     |> HTTPoison.get(rest_headers())
     |> case do
       {:ok, %{status_code: 200, body: body}} ->
-        map_contributors_response(body)
+        map_contributors(body)
 
       _ ->
         :timer.sleep(retry_delay())
         fetch_contributors(repo, try_count + 1)
     end
   end
 
-  defp map_contributors_response(body) do
-    body
-    |> Poison.decode()
-    |> map_contributors()
-  end
-
-  defp map_contributors({:ok, contributors}) when is_list(contributors) do
-    contributors
-    |> Enum.map(fn
-      %{"author" => %{"login" => username}} = contributor ->
-        {total, recent_commit_count, relevancy} = Contributors.calculate_relevancy(contributor)
-
-        %Contributor{username: username, relevancy: relevancy, recent_commit_count: recent_commit_count, total_commit_count: total}
+  defp map_contributors(body) do
+    case Jason.decode(body) do
```

Comment:
I think we use the `foo |> case do` pattern in the code base.

```elixir
body
|> Jason.decode()
|> case do
…
```

Reply:
> 3 lines of code to express logic that belongs together

I’m not sure I get what you’re saying 🤔 How is

```elixir
body
|> Jason.decode()
|> case do
```

different than

```elixir
items
|> Enum.map(& String.upcase(&1))
|> Enum.reject(& &1 == "FOO")
```

?

Reply:
Both snippet do express the same intention :

```elixir
1   body
2  |> Jason.decode()
3  |> case do
4    …
5  end
```

```elixir
1  case Jason.decode(body) do
2    …
3  end
```

But I think the later is more explicit. 

The `|>` operator, even though ideal to express data flow, tend to be overused. Which causes the code to be clever/complex for now valuable reasons. The same way pattern matching also tend to be overused instead of simple function parameters.

Reply:
@simonprev thoughts?

Reply:
We can’t really compare with your second example that would expand to and become unreadable!

```elixir
Enum.reject(Enum.map(tems, & String.upcase(&1)), & &1 == "FOO")
```

Reply:
I think we need to keep de "full pipe" if we want to be constant.

If you want to keep your intention and be explicit you can always do:

```elixir
decoded_body = Jason.decode(body)

case decoded_body do
  …
end
```

The less edge cases we have the easier it is for newcomers

Reply:
🤔 I reverted it back to piping, but we’ll have to talk about this again…

Reply:
> your second example that would expand to and become unreadable!

That’s the problem, that’s subjective 😉 Who decides what is readable for

```elixir
case MySuperLongModule.with_a_extra_long_function(having_many, params) do
end
```

```elixir
having_many
|> MySuperLongModule.with_a_extra_long_function(params)
|> case do
```

Reply:
@simonprev Stop being so _java-ish_! 🤣 

I still believe for this specific case, piping a value to a single function then to a "terminal" case statement is _clever_; even with_an_extra_long_function_name… Since this discussion is purely subjective,  let's agree to disagree. 😉